### PR TITLE
Start with parallel processing

### PR DESCRIPTION
**Pull request by anikaweinmann**
- Parallel Ephemeral Processing
  - Importer
  - Stdout

**TODO in other PR:**
- Parallel Persistent Processing
- Parallel Ephemeral Processing - Exporter
- Parallel Ephemeral Processing - reuse of exporter in next PC

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,6 +30,15 @@ docker-compose -f docker/docker-compose.yml run --rm --service-ports --entrypoin
 (cd /src/actinia-parallel-plugin && python3 setup.py install)
 # start actinia-core with your plugin
 gunicorn -b 0.0.0.0:8088 -w 1 --access-logfile=- -k gthread actinia_core.main:flask_app
+
+# or for debugging in one line with reset
+reset && (cd /src/actinia-parallel-plugin && python3 setup.py install) && gunicorn -b 0.0.0.0:8088 -w 3 --access-logfile=- -k gthread actinia_core.main:flask_app
+```
+
+### Postgis
+Connect to postgis DB from actinia-core docker container:
```
Comment (neteler) [inline README.md:39]:
```suggestion
### PostGIS
Connect to PostGIS DB from actinia-core docker container:
```

File: README.md [hunk 2]
```diff
@@ -77,4 +78,43 @@ make integrationtest
...
+* using stdout/export in PC of next block
+
+#### Persistent processing
+You can also start a **persistent** batch job via:
+```
+# parallel persistent processing (TODO!!!)
+curl -u actinia-gdi:actinia-gdi -X POST -H 'Content-Type: application/json' -d @test_postbodies/parallel_processing.json http://localhost:8088/api/v3/locations/nc_spm_08_grass7_root/mapsets/test_mapset/processing_parallel | jq
+```
+Hereby, the parallel started process chains will be computed in mapsets with
+the suffix `_parallel_{NUMBER}` (see teh example process chains in
```
Comment (neteler) [inline README.md:115]:
```suggestion
the suffix `_parallel_{NUMBER}` (see the example process chains in
```

File: setup.cfg [hunk 3]
```diff
@@ -7,7 +7,7 @@ name = actinia_parallel_plugin.wsgi
 description = actinia example plugin
 author = Anika Weinmann
 author-email = aweinmann@mundialis.de
-license = mit
+license = GPU3
```
Comment (neteler) [inline setup.cfg:10]:
```suggestion
license = GPL3
```

File: src/actinia_parallel_plugin/api/job.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+from actinia_parallel_plugin.resources.logging import log
+from actinia_parallel_plugin.core.jobs import (
+    getJob,
+)
+from actinia_parallel_plugin.apidocs import jobs
+
+
+class JobId(Resource):
+    """ Definition for endpoint standortsicherung
+    @app.route('/processes/standortsicherung/jobs/<jobid>')
```
Comment (neteler) [inline src/actinia_parallel_plugin/api/job.py:43]:
```suggestion
    """ Definition for endpoint standortsicherung
    @app.route('/processes/standortsicherung/jobs/<jobid>')
```

I assume that "standortsicherung" isn't an ideal word in this class...

File: src/actinia_parallel_plugin/model/batch.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            'items': {'type': 'object'}
+        },
+        'actinia_core_url': {
+            'type': 'string',
+            'description': ('The actinia-core IP or URL where actinia-core '
+                            'is processing the batchjob')
+        },
+        'actinia_gdi_batchid': {
+            'type': 'integer',
+            'description': 'The batch ID'
```
Comment (mmacata) [inline src/actinia_parallel_plugin/model/batch.py:207]: same here - we don't really need all of them

File: src/actinia_parallel_plugin/model/batch_process_chain.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    job as well.
+    """
+    version = fields.StringField()  # string
+    list = fields.ListField([Module], required=True)  # array of objects
+    # the following are given by the user to the BatchProcessChain and are
+    # then internally applied to each SingleJob
+    processing_platform = fields.StringField()  # string
+    processing_platform_name = fields.StringField()  # string
+    processing_host = fields.StringField()  # string
+    # processing_block, batch_description and
```
Comment (mmacata) [inline src/actinia_parallel_plugin/model/batch_process_chain.py:124]: and once more, some of these fields might not be needed.

File: src/actinia_parallel_plugin/model/jobtabelle.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    status = CharField(null=True)
+    actinia_core_response = BinaryJSONField(null=True)
+    actinia_core_jobid = CharField(null=True)
+    actinia_core_url = CharField(null=True)
+    actinia_core_platform = CharField(null=True)
+    actinia_core_platform_name = CharField(null=True)
+    terraformer_id = IntegerField(null=True)
+    terraformer_response = BinaryJSONField(null=True)
+    creation_uuid = CharField(null=True)
+    message = CharField(null=True)
```
Comment (mmacata) [inline src/actinia_parallel_plugin/model/jobtabelle.py:83]: and here as well. We might get rid of some of these as they are not needed in this context.

File: src/actinia_parallel_plugin/model/response_models.py [hunk 8]
```diff
@@ -49,3 +49,36 @@ class SimpleStatusCodeResponseModel(Schema):
...
+            'type': 'string',
+            'description': 'The coordinate reference system of the result.'
+        },
+        'table': {
+            'type': 'string',
+            'description': ('The database connection string of the source ' +
+                            'of the result.')
+        }
+    }
+    required = ["uuid", "bbox"]
```
Comment (mmacata) [inline src/actinia_parallel_plugin/model/response_models.py:84]: Is this used?

**Review by mmacata — COMMENTED** [review]
Nice! Thanks a lot!
I made some review comments and additionally some general thoughts:
- for the database table fields and models, some attributes might not be needed
- It would be great to get rid off all `Regeldatei` / `Rulefile` occurances, as this schema is not used here
- Please re-check the copyright notes. I found `Copyright (c) 2018-present` at least once and also inconsistencies between the two header years
- I like the endpoint `/api/v3/locations/nc_spm_08_grass7_root/processing_parallel` but I am not so sure about the "general" ones with `/api/v3/processing_parallel/batchjobs/1` and `/api/v3/processing_parallel/jobs/1`. It might be a bit confusing as usually the jobs can be found at the according ressource within actinia-core. I would like to discuss some alternatives here, maybe next week in person ;)

File: docker/docker-compose.yml [hunk 9]
```diff
@@ -7,10 +7,14 @@ services:
       dockerfile: docker/Dockerfile
     volumes:
       - ..:/src/actinia-parallel-plugin/.
+      # - /home/.../grassdata/:/actinia_core/grassdb/
```
[the remark this answers is not in the document]
Comment (anikaweinmann) [reply docker/docker-compose.yml:10]: I used it locally with my grassdata folder

File: src/actinia_parallel_plugin/apidocs/jobs.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+script_dir = os.path.dirname(os.path.abspath(__file__))
+
+rel_path = "../apidocs/examples/jobs_get_docs_response_example.json"
+abs_file_path = os.path.join(script_dir, rel_path)
+print(abs_file_path)
+with open(abs_file_path) as jsonfile:
+    jobs_get_docs_response_example = json.load(jsonfile)
+
+
+class EnrichedRegeldateiModel(Schema):
```
[the remark this answers is not in the document]
Comment (anikaweinmann) [reply src/actinia_parallel_plugin/apidocs/jobs.py:44]: How would you name it?
`Batchfile` ?

File: src/actinia_parallel_plugin/apidocs/jobs.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+                           'idpk_jobs is known. (More unique than creation '
+                           'timestamp)'
+        },
+        'terraformer_id': {
+            'type': 'string',
+            'description': 'The terraformer instances ID for the job'
+        },
+        'terraformer_response': {
+            'type': 'string',
+            'description': 'The Response/Status of terraformer'
```
[the remark this answers is not in the document]
Comment (anikaweinmann) [reply src/actinia_parallel_plugin/apidocs/jobs.py:161]: Don't we need this anymore? I can delete everything that is not needed at the moment, but I thought we would need at least half of it again soon.

**Comment by anikaweinmann** [comment]
> * I like the endpoint `/api/v3/locations/nc_spm_08_grass7_root/processing_parallel` but I am not so sure about the "general" ones with `/api/v3/processing_parallel/batchjobs/1` and `/api/v3/processing_parallel/jobs/1`. It might be a bit confusing as usually the jobs can be found at the according ressource within actinia-core. I would like to discuss some alternatives here, maybe next week in person ;)

- `/api/v3/resouces/USERID/batches/1` (Only users who are in the group that created the job are allowed to query it!)
- `/api/v3/resouces/USERID/batches/1/jobs/1` (Only users who are in the group that created the job are allowed to query it!)
- `/api/v3/resouces/USERID/jobs/resource_id_without_"resource_id"` 


   - Equal to `/api/v3/resouces/USERID/resource_id_with_"resource_id"` (will be done later in actinia_core)
   - for iterations all are returned
   - `/api/v3/resouces/USERID/jobs/resource_id_without_"resource_id"/iterations"/ITERATION_NUMBER` returns only one iteration
   - `/api/v3/resouces/USERID/jobs/resource_id_without_"resource_id"/iterations/latest` returns only the last iteration

File: src/actinia_parallel_plugin/model/batch.py (hunk 5 above)
Comment (anikaweinmann) [reply src/actinia_parallel_plugin/model/batch.py:207]: `actinia_gdi_batchid` --> `batchid`

File: src/actinia_parallel_plugin/apidocs/jobs.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (anikaweinmann) [reply src/actinia_parallel_plugin/apidocs/jobs.py:160]: I will move it

**Comment by anikaweinmann** [comment]
@mmacata the endpoints are changed as discussed, please review again.
Dependencies are added to README in https://github.com/mundialis/actinia_core/pull/356

File: src/actinia_parallel_plugin/wsgi.py [hunk 12]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env python
```
Comment (mmacata) [inline src/actinia_parallel_plugin/wsgi.py:1]: I think this file is not needed - only needed when this flask api should run as separate api from actinia_core

File: src/actinia_parallel_plugin/core/ephemeral_processing.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Copyright (c) 2022 mundialis GmbH & Co. KG
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
```
[the remark this answers is not in the document]
Comment (anikaweinmann) [reply src/actinia_parallel_plugin/core/ephemeral_processing.py:7]: will be done in MR for parallel processing

**Review by mmacata — APPROVED** [review]
Thanks a lot! Great :)

**Pull request merged**